File: backend/api/contracts/portfolio.go
package contracts

import (
	"github.com/google/uuid"
)

type Portfolio struct {
	ID    uuid.UUID           `json:"id"`
	Name  string              `json:"name"`
	Items []PortfolioListItem `json:"items"`
}

type PortfolioListItem struct {
	ID     uuid.UUID `json:"id"`
	FundID uuid.UUID `json:"fundId"`
	Name   string    `json:"name"`
	Amount float64   `json:"amount"`
}

type PortfolioDetails struct {
	FundInformation               []FundInformation               `json:"fundInformation"`
	Sectors                       []string                        `json:"sectors"`
	PortfolioFundSectorWeightings []PortfolioFundSectorWeightings `json:"portfolioFundSectorWeightings"`
}

type PortfolioFundSectorWeightings struct {
	FundName            string                `json:"fundName"`
	PercentageOfTotal   float64               `json:"percentageOfTotal"`
	FundSectorWeighting []FundSectorWeighting `json:"fundSectorWeighting"`
}

File: ui/src/lib/portfolio.ts
import type {FundInformation, FundSectorWeighting} from "$lib/fund";
import type {PortfolioFundSectorWeighting__Output} from "$lib/proto/proto/PortfolioFundSectorWeighting";

export interface Portfolio {
    id: string;
    name: string;
    items: PortfolioListItem[];
}

export interface PortfolioListItem {
    id: string;
    fundId: string;
    name: string;
    amount: number | undefined;
}

export interface PortfolioDetails {
    fundInformation: FundInformation[]
    sectors: string[]
    portfolioFundSectorWeightings: PortfolioFundSectorWeightings[]
}

export interface PortfolioFundSectorWeightings {
    fundName: string;
    percentageOfTotal: number;
    fundSectorWeighting: FundSectorWeighting[]
}

export interface PortfolioHoldingsFilter {
    portfolioId: string;
    searchTerm: string;
    sectorName: string;
    limit: number;
    offset: number;
}

export interface PortfolioSectorWeighting {
    sectorName: string
    weighting: PortfolioFundSectorWeighting__Output
}